samples/web_shell/scripts/git.js

```diff
@@ -0,0 +1,319 @@
+    message['subject'] = genMessageId();
+    message.arg = args;
+    return message;
+  });
+}
+
+function createMessageCallback(cx, message) {
+  var cb = function(result) {
+    console.log(message['name'] + ' command successful!');
+    cx.closeOk();
+  };
+  return cb;
+}
+
+/**
+ *@constructor
+ */
+var GitSalt = function() {
+  this.callbacks = new Object();
+};
+
+GitSalt.prototype.listenerDiv = null;
+
+GitSalt.prototype.loadPromise = null;
+
+GitSalt.prototype.callbacks = {};
+
+/**
+ * Create the Native Client <embed> element as a child of the DOM element
+ * named "listener".
+ *
+ * @param {string} name The name of the example.
+ * @param {string} path Directory name where .nmf file can be found.
+ */
+GitSalt.prototype.createNaClModule = function(name, path) {
+  return new Promise(function(resolve, reject) {
+    this.loadPromise = resolve;
+    var moduleEl = document.createElement('embed');
+    moduleEl.setAttribute('width', 0);
+    moduleEl.setAttribute('height', 0);
+    moduleEl.setAttribute('path', path);
+    moduleEl.setAttribute('src', path + '/' + name + '.nmf');
+
+    moduleEl.setAttribute('type', "application/x-pnacl");
+    this.naclModule = moduleEl;
+
+    // The <EMBED> element is wrapped inside a <DIV>, which has both a 'load'
+    // and a 'message' event listener attached.  This wrapping method is used
+    // instead of attaching the event listeners directly to the <EMBED> element
+    // to ensure that the listeners are active before the NaCl module 'load'
+    // event fires.
+    this.listenerDiv.appendChild(moduleEl);
+  }.bind(this));
+};
+
+GitSalt.prototype.statusText = 'NO-STATUSES';
+
+GitSalt.prototype.updateStatus = function(opt_message) {
+  if (opt_message) {
+    statusText = opt_message;
+  }
+}
+
+/**
+ * Add the default "load" and "message" event listeners to the element with
+ * id "listener".
+ *
+ * The "load" event is sent when the module is successfully loaded. The
+ * "message" event is sent when the naclModule posts a message using
+ * PPB_Messaging.PostMessage() (in C) or pp::Instance().PostMessage() (in
+ * C++).
+ */
+GitSalt.prototype.attachDefaultListeners = function() {
+  this.listenerDiv.addEventListener('load', this.moduleDidLoad.bind(this), true);
+  this.listenerDiv.addEventListener('message', this.handleMessage.bind(this), true);
+  this.listenerDiv.addEventListener('error', this.handleError.bind(this), true);
+  this.listenerDiv.addEventListener('crash', this.handleCrash.bind(this), true);
+};
+
+/**
+ * Called when the NaCl module fails to load.
+ *
+ * This event listener is registered in createNaClModule above.
+ */
+GitSalt.prototype.handleError = function(event) {
+  this.updateStatus('ERROR [' + this.naclModule.lastError + ']');
+};
+
+/**
+ * Called when the Browser can not communicate with the Module
+ *
+ * This event listener is registered in attachDefaultListeners above.
+ */
+GitSalt.prototype.handleCrash = function(event) {
```

Comment:
Shouldn't we propagate (somehow) the error to cx so we display an error in the wash window?